GATKTool pre/post ReadTransformers methods + ReadWalker/LocusWalker integration

magicDGS opened the pull request:
In the same way as `makeReadFilter()`, I implemented a function to include read transformers in `ReadWalker` and `LocusWalker` to use already transformed reads in `apply` (as requested in #1886).

This function could be added to the `GATKSparkTool` too, and include it in `getReads()`.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java:
```diff
@@ -121,8 +136,12 @@ public void traverse() {
+        final Iterator<GATKRead> readIterator = StreamSupport.stream(reads.spliterator(), false).map(transformer).iterator();
         // get the LIBS
-        LocusIteratorByState libs = new LocusIteratorByState(new ReadFilteringIterator(reads.iterator(), countedFilter), getDownsamplingMethod(), includeDeletions(), includeNs(), keepUniqueReadListInLibs(), samples, header);
+        LocusIteratorByState libs = new LocusIteratorByState(new ReadFilteringIterator(readIterator, countedFilter), getDownsamplingMethod(), includeDeletions(), includeNs(), keepUniqueReadListInLibs(), samples, header);
```
Can you document in the class-level docs for `LocusWalker` that read transformers are applied *before* read filters?

magicDGS replied:
Documented in the class javadoc.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -72,9 +73,11 @@ public void traverse() {
 
         StreamSupport.stream(reads.spliterator(), false)
                 .filter(countedFilter)
+                .map(transformer)
```
Here in `ReadWalker` you are applying the transformers after the filters, which is the opposite of the behavior in `LocusWalker`. The behavior should definitely be consistent across the two walker types. We think that the transformers should be applied before the filters in both cases, so that you have an opportunity to repair bad input that would otherwise be filtered out.

magicDGS replied:
Sure, it was an error. I will do it in the next commit! Thanks!

magicDGS replied:
Keep it as now (filter/transformer). I guess that that's why I change it here, because probably the tests were failing.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSR.java:
```diff
@@ -47,18 +47,23 @@
     @Override
     public void apply( GATKRead read, ReferenceContext referenceContext, FeatureContext featureContext ) {
-        outputWriter.addRead(transform.apply(read));
+        outputWriter.addRead(read);
```
I think that `ApplyBQSR` should continue to apply the read transformer manually, rather than via the new `GATKTool` interface, so that it can guarantee that it's applied only to filtered reads (since it could fail on malformed reads). 

There's a bit of an issue here with some read transformers that definitely should be applied before filtering (like `FixMisencodedBaseQualityReads`) to give an opportunity to repair bad input that would otherwise be filtered, and others (like `BQSRReadTransformer`) that should definitely be applied after filtering so that they don't fail on malformed input. We may eventually need a mechanism to specify whether a given transformer should be applied before or after filtering. Thoughts on this?

lbergelson replied:
We could just leave bqsr to apply the transformer explicitly call transform manually during apply.

magicDGS replied:
Because I kept it with filter/transform, this depends on the final decision.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/rnaseq/SplitNCigarReads.java:
```diff
@@ -119,7 +126,7 @@ public void onTraversalStart() {
     @Override
     public void apply(GATKRead read, ReferenceContext referenceContext, FeatureContext featureContext) {
-        splitNCigarRead(rnaReadTransform.apply(read),overhangManager);
+        splitNCigarRead(read, overhangManager);
```
Argh, so it seems like `NDNCigarReadTransformer` is another one that wants to go after read filtering rather than before, since it blows up given an invalid cigar (which would be caught by the default read filters). So out of our three current read filters, two (`NDNCigarReadTransformer` and `BQSRReadTransformer`) want to be applied after filtering, and one (`FixMisencodedBaseQualityReads`) wants to go before filtering. @lbergelson @cmnbroad can you comment on this issue?

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java:
```diff
@@ -6,11 +6,13 @@
+import org.broadinstitute.hellbender.utils.read.GATKRead;
 
 import java.util.*;
 import java.util.stream.Collectors;
```
Can you create a `LocusWalkerUnitTest` and `ReadWalkerUnitTest`, and add basic tests for the new read transformer functionality to those two classes?

magicDGS replied:
Done

droazen commented:
Review complete, thanks for this contribution! It looks good, but there is an ordering issue of read transformers vs. read filters that we're going to have to discuss and come to a decision on before this is merged.

magicDGS commented:
Thanks for the review, @droazen. My ideas about the ordering are the following: in #2084, I have the implementation of the fixing/checking of misencoded quals included in the read sources. So the only one that could be applied before filtering is gone if you accept that PR.

Thus, I guess that the transformer could be applied after the filtering. Because transformers could be kind of demanding in computation, it will be interesting to pass directly to them filtered reads. In addition, could be included in the contract that only valid reads for the tool ensure that the read transformers don't blow up.

Whatever the decision is, should I wait for addressing the comments?

magicDGS commented:
I updated the branch with the new changes in the framework, and addressing the comments. Back to you, @droazen.

droazen commented:
I think there probably needs to be a mechanism in the code to control whether each transformer is applied before or after filtering. Transformers meant to repair bad input need to go before, and those meant to process it in some other way probably need to go after. @lbergelson @cmnbroad @magicDGS suggestions on what form this should take?

magicDGS commented:
The first idea that comes to my mind is using to methods: `makeReadTransformerBefore` and `makeReadTransformerAfter`, and implementations could control the order of the transformers. Other idea is create two interfaces for transform reads: `ReadRepairTransformer` and `ReadProcessingTransformer`, and make the two methods return the specific interface to do not allow one or the other to be applied before/after.

cmnbroad commented:
makeBefore/makeAfter seems more flexible in that it makes before/after a property of the way the transformer is applied, rather than of the transformer itself. When we write the plugin descriptor, it can maintain two separate argument lists (for "--preFilterTransformer ..." and "--postFilterTransformer ..."), and then merge them accordingly. It does complicate the tool structure, but its the price for flexibility, and the pattern is not that complex.

I think we should be explicit about what "pre" and "post" are relative to in the method and argument names, i.e., makePreReadFilterTransformer and makePostReadFilterTransformer, or maybe even one makeReadFilterTransformer method that takes a pre/post argument.

droazen commented:
This problem with this idea is that `ReadTransformers` will be able to be turned on/off by the user. So if a user requests that the `FixMisencodedBaseQualityReads` be turned on, the engine needs to make a decision about whether to apply it before or after the filters.

cmnbroad commented:
I don't see why; the user would disambiguate it by saying either "--preReadFilterTransformer FixMisencodedBaseQualityReads" "--postReadFilterTransformer FixMisencodedBaseQualityReads".

droazen commented:
Ah I see -- yes, that could work.

lbergelson commented:
`--preReadFilterTransformer FixMisencodedBaseQualityReads` <- why we need commandline auto completion

lbergelson commented:
We could combine the filters and transformers into 1 plugin that can intersperse them.  It would be more complicated but be maximally expressive.

droazen commented:
I'll take the long version over some cryptic abbreviation like `--PRFT`  :)

magicDGS commented:
I implemented the pre/post methods and added tests for them, @droazen. I guess that this is ready for review.

droazen commented:
Re-assigning to @cmnbroad to review the pre/post transformer changes.

magicDGS commented:
Rebased to current master, @cmnbroad.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java:
```diff
@@ -26,6 +26,9 @@
  * one or multiple sources of reads, with optional contextual information from a reference and/or sets of
  * variants/Features.
  *
+ * Reads will be filter with {@link #makeReadFilter()} and transformed with {@link #makePostReadFilterTransformer()} before include
```
"will be filter" -> "will be filtered"

magicDGS replied:
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/LocusWalker.java:
```diff
@@ -137,6 +140,32 @@ protected final LIBSDownsamplingInfo getDownsamplingInfo() {
     }
 
     /**
+     * Returns the pre-filtered read transformer (simple or composite) that will be applied to the reads before calling {@link #apply}.
```
"Returns the pre-filtered" -> "Returns the pre-filter"

magicDGS replied:
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -18,6 +19,9 @@
  * If multiple sources of reads are specified, they are merged together into a single sorted stream of reads.
  *
+ * Reads will be filter with {@link #makeReadFilter()} and transformed with {@link #makePostReadFilterTransformer()} before pass
+ * them to apply().
```
This comment needs updating - I think it was from before we had pre and post transformers.

magicDGS replied:
Done.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/ApplyBQSR.java:
```diff
@@ -47,18 +47,23 @@
 
-    private ReadTransformer transform;
+    /**
+     * Returns the BQSR transformer
```
BQSR transformer -> BQSR post transformer.

magicDGS replied:
Done.

magicDGS commented:
I updated the javadoc with your comments @cmnbroad. Should I only rebase or also squash my previous commits? Waiting for your answer to continue working on this...

cmnbroad commented:
Go ahead and squash as well. thx.

magicDGS commented:
Now it is rebased/squashed. Back to you for the next round, @cmnbroad.

cmnbroad commented:
@magicDGS Can you move the make*Transformer methods up to GATKTool (at some point we'll have a plugin at that level), and then also integrate these with AssemblyRegionWalker ? We'll want to do the Spark tools as well, but we leave that for a separate PR.

magicDGS commented:
I moved the transformers up to GATK tool in the last commit (the previous one was fixing the barclay change in the tests).

Integraction with `AssemblyRegionWalker` is more complicated, because it requires to modify the read shard implementation to include transformers there. Maybe we should open an issue regarding that, or I can include a setter for pre/post transformers in the changes for #2372.

Back to you @cmnbroad.

droazen commented:
@magicDGS Sorry for the delay on this one. We do want to get this merged soon, so that we can take the next step of making these pluggable. @cmnbroad and I had a quick glance at this today, and it looks good now, just a few little changes required. @cmnbroad will add his comments to this PR shortly.

magicDGS commented:
Thanks for the information @droazen. I will do the changes as soon as @cmnbroad comment it!

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/ReadWalkerUnitTest.java:
```diff
@@ -0,0 +1,81 @@
+        Assert.assertEquals(tool.totalReads, 5);
+    }
+
+}
```
I think we need more comprehensive validation in both of these sets of tests (LocusWalker and ReadWalker). We should use a transformer that actually does a transformation, and also validates that the pipeline (pre-transform->filter->post-transform) is executed in the correct order. As it stands, if we were to introduce a regression that changed the execution order of pre, filter, post, I think these would still pass.

magicDGS replied:
I modify slightly the tests to assess that the pre-transform -> filter -> post-transform order is correctly executed, in both `LocusWalker` and `ReadWalker`

magicDGS commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -117,6 +123,33 @@ SimpleInterval getReadInterval(final GATKRead read) {
+    }
+
+
+    /**
```
Done.

magicDGS commented:
Back to you @cmnbroad, I think that this is almost there!

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/engine/LocusWalkerUnitTest.java:
```diff
@@ -0,0 +1,109 @@
+        @Override
+        public ReadTransformer makePreReadFilterTransformer() {
+
+            // transforming name of unmapped mates
```
I think these comments are out-of-sync with the code ?

magicDGS replied:
True. Fixed.

cmnbroad commented:
Thanks @magicDGS !
